File: RB101_RB109_small_problems/easy_8/palindromic_substrings.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: string
-  Output: array of all substrings that are palindromic

---

**Problem Domain:**
Palindromes are strings that are the same forwards and backwards.
---

**Implicit Requirements:**
- Use #substrings method from previous exercise.
- Return values should be arranges in same sequence as the substrings appear.
- Duplicate palindromes should be maintained.
- Palindromes are case specific.
- Single characters are not palindromes.
---

**Clarifying Questions:**

1.

---

**Mental Model:**

- Generate all substrings of the input string.
- Select which substrings are the same forwards as backwards (or reversed).

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

palindromes('abcd') == []
palindromes('madam') == ['madam', 'ada']
palindromes('hello-madam-did-madam-goodbye') == [
  'll', '-madam-', '-madam-did-madam-', 'madam', 'madam-did-madam', 'ada',
  'adam-did-mada', 'dam-did-mad', 'am-did-ma', 'm-did-m', '-did-', 'did',
  '-madam-', 'madam', 'ada', 'oo'
]
palindromes('knitting cassettes') == [
  'nittin', 'itti', 'tt', 'ss', 'settes', 'ette', 'tt'
]

---

Data Structure
--------------

- Use arrays to store substrings and iterate over for selection.

---

Algorithm
---------

- Get all substrings within input string using previous exercises #substrings method.
- Perform selection on the array of substrings
  - Only select those substrings that are greater than 1 character and the same forwards as backwards
- Return selection array.

Code
----

=end

def leading_substrings(string)
  result = []
  0.upto(string.size - 1) do |index|
    result << string[0..index]
  end
  result
end

def substrings(string)
  all_substrings = []
  string.length.times do |index|
    all_substrings += leading_substrings(string[index..-1])
  end
  all_substrings
end

def palindrome?(string)
  (string.length > 1) && (string == string.reverse)
end

def palindromes(string)
  substrings(string).select do |substring|
    palindrome?(substring)
  end
end

#Further Exploration - ignore non-alphanumeric characters and case.

def remove_non_alphanumeric(string)
  string.delete("^0-9A-Za-z")
end

def palindrome?(string)
  string = remove_non_alphanumeric(string).downcase
  (string.length > 1) && (string == string.reverse)
end

File: JS210_small_problems/list_processing/palindromic_substrings.js
function leadingSubstrings(str) {
  let substrings = [];

  for (let length = 1; length <= str.length; length += 1) {
    substrings.push(str.slice(0, length));
  }

  return substrings;
}

function substrings(str) {
  let substringsArray = [];
  for (let index = 0; index < str.length; index += 1) {
    let partialStr = str.slice(index);
    substringsArray = substringsArray.concat(leadingSubstrings(partialStr));
  }

  return substringsArray;
}

function isPalindrome(str) {
  return (str.length > 1) && (str === str.split('').reverse().join(''));
}

function palindromes(str) {
  return substrings(str).filter(isPalindrome);
}

console.log(palindromes('abcd'));       // []
console.log(palindromes('madam'));      // [ "madam", "ada" ]

console.log(palindromes('hello-madam-did-madam-goodbye'));
// // returns
// // [ "ll", "-madam-", "-madam-did-madam-", "madam", "madam-did-madam", "ada",
// //   "adam-did-mada", "dam-did-mad", "am-did-ma", "m-did-m", "-did-", "did",
// //   "-madam-", "madam", "ada", "oo" ]

console.log(palindromes('knitting cassettes'));
// // returns
// // [ "nittin", "itti", "tt", "ss", "settes", "ette", "tt" ]